packages/docusaurus-theme-common/src/utils/useTOCHighlight.ts

```diff
@@ -25,19 +29,30 @@ function isInViewportTopHalf(boundingRect: DOMRect) {
   return boundingRect.top > 0 && boundingRect.bottom < window.innerHeight / 2;
 }
 
-function getAnchors() {
-  // For toc highlighting, we only consider h2/h3 anchors
-  const selector = '.anchor.anchor__h2, .anchor.anchor__h3';
+function getAnchors({
+  minHeadingLevel,
+  maxHeadingLevel,
+}: {
+  minHeadingLevel: number;
+  maxHeadingLevel: number;
+}) {
+  const selectors = [];
+  for (let i = minHeadingLevel; i <= maxHeadingLevel; i += 1) {
+    selectors.push(`.anchor.anchor__h${i}`);
```

Comment:
I'm not sure to understand what you mean here.

We want to get only the anchors from min to max, so we need levels in classnames otherwise we'd select more anchors and it could mess-up with the TOC algorithm (an anchor that is not in the TOC should never be considered as "active" because we wouldn't be able to highlight it)

Reply:
>  could mess-up with the TOC algorithm 

How can something break if we get all anchor links? If it important, we can actually filter out anchor links based on its heading tag instead class.

Reply:
It is important. Consider this scenario:
- We config to only use h2 in toc (not h3)
- We scroll to a h3 heading

Then, the h3 heading is considered the "active" one but it's not in the TOC, so nothing is highlighted (while we should highlight the parent h2)

![image](https://user-images.githubusercontent.com/749374/135266156-a941aee2-e2b5-4875-8c04-54324f8cffbc.png)

--- 

> we can actually filter out anchor links based on its heading tag instead class.

You can do `select("h2 h3 h4")` or `select("*").filter(is("h2 h3 h4"))`, both would lead to the same result

It's a best practice to create a more precise selector / query in the first place when possible

Reply:
>  both would lead to the same result

Yes, but we can get rid of extra classes, why do we need them if can avoid using them?

Reply:
It doesn't seem like a big deal to me to keep them and would be a CSS breaking change to remove them now

Also nothing guarantees that a user won't use headings in other places of its theme for weird reasons, and we may want to avoid that it conflicts with the TOC algorithm by not using `"h2 h3"` selector on purpose. Classnames help remove the opportunities for such a conflict

Reply:
We've already marked this PR as BC, so it's good time to get rid of this extra thing.

Right, but our header-anchors will still have  generic class `anchor`, though  could make it even more specific. I suggest removing only heading-specific classes, and use heading tag level instead in TOC algorithm.

Reply:
Oh I thought you wanted to remove both classes

If you want to refactor to `h2.anchor, h3.anchor`, I'm fine with the idea (and `.anchor__h2` is quite new so it's a smaller breaking change)